src/evaluators/BinaryExpression.js

```diff
@@ -189,30 +173,81 @@ export function computeBinary(
     }
   }
 
-  if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
-    // generate error if binary operation might throw or have side effects
-    if (realm.isInPureScope()) {
-      realm.evaluateWithPossibleThrowCompletion(
-        () => {
-          getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
-          return realm.intrinsics.undefined;
-        },
-        TypesDomain.topVal,
-        ValuesDomain.topVal
-      );
-    } else {
-      getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
+  const previousErrorHandler = realm.errorHandler;
+
+  let isPure = true;
+  if (realm.isInPureScope()) {
+    // If we're in pure mode we can recover even if this operation might not be pure.
+    // To do that, we'll temporarily override the error handler.
+    realm.errorHandler = diagnostic => {
+      isPure = false;
+      return "Recover";
+    };
+  }
+
+  let resultType;
+  let effects;
+  try {
+    effects = realm.evaluateForEffects(() => {
+      if (lval instanceof AbstractValue || rval instanceof AbstractValue) {
+        // generate error if binary operation might throw or have side effects
+        resultType = getPureBinaryOperationResultType(realm, op, lval, rval, lloc, rloc);
```

Comment:
This code gets complicated by trying to retain this resultType. If I can instead assume that this type will always be the same as `AbstractValue.createFromBinaryOp(...).getType()` below, I can simplify this code. I think that should be correct.